packages/medusa/src/api/routes/admin/products/create-variant.ts

```diff
@@ -105,58 +119,126 @@ import { defaultFields, defaultRelations } from "./"
-    manage_inventory: Validator.boolean().optional(),
-    weight: Validator.number().allow(null).optional(),
-    length: Validator.number().allow(null).optional(),
-    height: Validator.number().allow(null).optional(),
-    width: Validator.number().allow(null).optional(),
-    origin_country: Validator.string().allow(""),
-    mid_code: Validator.string().allow(""),
-    material: Validator.string().allow(""),
-    metadata: Validator.object().optional(),
-    prices: Validator.array()
-      .items(
-        Validator.object()
-          .keys({
-            region_id: Validator.string().empty(null),
-            currency_code: Validator.string().required(),
-            amount: Validator.number().integer().required(),
-            sale_amount: Validator.number().optional(),
-          })
-          .xor("region_id", "currency_code")
-      )
-      .required(),
-    options: Validator.array()
-      .items({
-        option_id: Validator.string().required(),
-        value: Validator.string().required(),
-      })
-      .default([]),
-  })
-
-  const { value, error } = schema.validate(req.body)
-  if (error) {
-    throw new MedusaError(MedusaError.Types.INVALID_DATA, error.details)
-  }
+  const validated = await validator(
+    AdminPostProductsProductVariantsReq,
+    req.body
+  )
 
-  const productVariantService = req.scope.resolve("productVariantService")
-  const productService = req.scope.resolve("productService")
+  const productVariantService: ProductVariantService = req.scope.resolve(
+    "productVariantService"
+  )
+  const productService: ProductService = req.scope.resolve("productService")
 
-  await productVariantService.create(id, value)
+  await productVariantService.create(id, validated)
 
   const product = await productService.retrieve(id, {
-    select: defaultFields,
-    relations: defaultRelations,
+    select: defaultAdminProductFields,
+    relations: defaultAdminProductRelations,
   })
 
   res.json({ product })
 }
+
+class ProductVariantOptionReq {
+  @IsString()
+  value: string
+
+  @IsString()
+  option_id: string
+}
+
+class ProductVariantPricesReq {
+  @Validate(XorConstraint, ["currency_code"])
+  region_id?: string
+
+  @Validate(XorConstraint, ["region_id"])
+  currency_code?: string
+
+  @IsNumber()
+  @Type(() => Number)
+  amount: number
+
+  @IsOptional()
+  @IsNumber()
+  @Type(() => Number)
+  sale_amount: number
+}
+
+export class AdminPostProductsProductVariantsReq {
+  @IsString()
+  title: string
+
+  @IsString()
+  @IsOptional()
+  sku?: string
+
+  @IsString()
+  @IsOptional()
+  ean?: string
+
+  @IsString()
+  @IsOptional()
+  upc?: string
+
+  @IsString()
+  @IsOptional()
+  barcode?: string
+
+  @IsString()
+  @IsOptional()
+  hs_code?: string
+
+  @IsNumber()
+  inventory_quantity = 0
+
+  @IsBoolean()
+  @IsOptional()
+  allow_backorder?: boolean
+
+  @IsBoolean()
+  @IsOptional()
+  manage_inventory?: boolean
+
+  @IsNumber()
+  @IsOptional()
+  weight?: number
```

Comment:
As this is a string number, shouldn't we use `@IsNumberString()` as is done [here](https://github.com/medusajs/medusa/pull/809)

Reply:
Or at least provide the transformer `@Type(() => Number)` ?

Reply:
Also valid for the other numbers below, as they are given as strings in the OAS description

Reply:
This is only needed when the numbers are being received as string in the `path`. In case of body payloads, we will receive json and can therefore safely use only Number :)